cvat/apps/engine/static/engine/js/shapes.js

```diff
@@ -1574,18 +1571,40 @@ class RaysController extends PolyShapeController {
         super(raysModel);
     }
 
+    static getDashArray(index) {
+        const dashes = {
+            0: [5, 0],
+            1: [5, 2, 1, 2],
+            2: [5, 1, 1, 1, 1, 1],
+        };
+        const auxiliary = [1, 1];
+        return (dashes[index] || auxiliary).map(v => v * DASH_LENGTH / window.cvat.player.geometry.scale).join(" ");
+    }
+
     updateLineCoordinates(lineElement, linePhi, mousePos) {
-        const { toPolarCoordinates, fromPolarCoordinates, rotate } = window.graphicPrimitives;
+        const {
+            toPolarCoordinates,
+            fromPolarCoordinates,
+            getOppositeAngle,
+            getAngleBetween,
+            rotate,
+        } = window.graphicPrimitives;
         const rotationPoint = this._model.vanishingPoint;
         const points = lineElement.array().value
             .map(([x, y]) => ({x, y}))
             .map(p => window.cvat.translate.box.canvasToActual(p));
         let newPoints;
+
         if (rotationPoint) {
-            const { phi } = toPolarCoordinates(mousePos, rotationPoint);
+            const { phi: fixedPhi } = toPolarCoordinates(mousePos, rotationPoint);
+            const revPhi = getOppositeAngle(fixedPhi);
+            function snapAngle({r, phi}) {
+                const newPhi = getAngleBetween(phi, fixedPhi) < getAngleBetween(phi, revPhi) ? fixedPhi : revPhi;
+                return {r, phi: newPhi};
+            }
             newPoints = points
                 .map(p => toPolarCoordinates(p, rotationPoint))
-                .map(coords => ({...coords, phi}))
+                .map(snapAngle)
```

Comment:
what is the point of this snapping?

Reply:
The goal is to project points onto the line which goes through rotation point (vanishing point) and current mouse position.